File: ts/earthbucks-lib/test/tx-builder.test.ts
import { describe, expect, test, beforeEach, it } from "vitest";
import { TxBuilder } from "../src/tx-builder.js";
import { TxOutBnMap } from "../src/tx-out-bn-map.js";
import { TxOut } from "../src/tx-out.js";
import { Script } from "../src/script.js";
import { KeyPair } from "../src/key-pair.js";
import { Pkh } from "../src/pkh.js";
import { PkhKeyMap } from "../src/pkh-key-map.js";
import { FixedBuf } from "@webbuf/fixedbuf";
import { TxOutBn } from "../src/tx-out-bn.js";
import { U8, U16BE, U32BE, U64BE } from "@webbuf/numbers";

describe("TxBuilder", () => {
  let txBuilder: TxBuilder;
  let txOutBnMap: TxOutBnMap;
  let pkhKeyMap: PkhKeyMap;

  beforeEach(() => {
    txOutBnMap = new TxOutBnMap();
    pkhKeyMap = new PkhKeyMap();
    // generate 5 keys, 5 outputs, and add them to the txOutMap
    for (let i = 0; i < 5; i++) {
      const key = KeyPair.fromRandom();
      const pkh = Pkh.fromPubKeyBuf(key.pubKey.toBuf());
      pkhKeyMap.add(key, pkh);
      const script = Script.fromPkhOutput(pkh);
      const txOut = new TxOut(new U64BE(100), script);
      const txOutBn = new TxOutBn(txOut, new U32BE(0n));
      txOutBnMap.add(txOutBn, FixedBuf.alloc(32), new U32BE(i));
    }

    const changeScript = Script.fromEmpty();
    txBuilder = new TxBuilder(txOutBnMap, changeScript, new U32BE(0n));
  });

  test("should build a valid tx when input is enough to cover the output", () => {
    const key = KeyPair.fromRandom();
    const pkh = Pkh.fromPubKeyBuf(key.pubKey.toBuf());
    const script = Script.fromPkhOutput(pkh);
    const output = new TxOut(new U64BE(50), script);
    txBuilder.addOutput(output);

    const tx = txBuilder.build();

    expect(tx.inputs.length).toBe(1);
    expect(tx.outputs.length).toBe(2);
    expect(tx.outputs[0]?.value.bn).toEqual(BigInt(50));
  });

  test("should build a valid tx when all inputs are enough to cover the output", () => {
    const key = KeyPair.fromRandom();
    const pkh = Pkh.fromPubKeyBuf(key.pubKey.toBuf());
    const script = Script.fromPkhOutput(pkh);
    const output = new TxOut(new U64BE(100 * 5 - 1), script);
    txBuilder.addOutput(output);

    const tx = txBuilder.build();

    expect(tx.inputs.length).toBe(5);
    expect(tx.outputs.length).toBe(2);
    expect(tx.outputs[0]?.value.bn).toEqual(BigInt(100 * 5 - 1));
    expect(tx.outputs[1]?.value.bn).toEqual(BigInt(1));
  });

  test("should build an invalid tx when input is insufficient to cover the output", () => {
    const txOut = new TxOut(new U64BE(10000), Script.fromEmpty());
    txBuilder.addOutput(txOut);

    const tx = txBuilder.build();

    expect(tx.inputs.length).toBe(5);
    expect(tx.outputs.length).toBe(1);
    expect(txBuilder.inputAmount.bn).toEqual(BigInt(500));
    expect(tx.outputs[0]?.value.bn).toEqual(BigInt(10000));
  });
});

File: rs/earthbucks_lib/tests/tx_builder_test.rs

use earthbucks_lib::key_pair::KeyPair;
use earthbucks_lib::pkh::Pkh;
use earthbucks_lib::script::Script;
use earthbucks_lib::tx_builder::*;
use earthbucks_lib::tx_out::*;
use earthbucks_lib::tx_out_bn_map::*;

fn setup() -> TxBuilder {
    let mut tx_out_bn_map = TxOutBnMap::new();
    let change_script = Script::from_iso_str("");

    for i in 0..5 {
        let key = KeyPair::from_random();
        let pkh = Pkh::from_pub_key_buffer(key.pub_key.buf.to_vec());
        let script = Script::from_pkh_output(pkh.to_iso_buf());
        let tx_out = TxOut::new(100, script);
        let block_num = 0;
        tx_out_bn_map.add(&[0; 32], i, tx_out, block_num);
    }

    TxBuilder::new(&tx_out_bn_map, change_script.unwrap(), 0)
}

#[test]
fn test_build_valid_tx_when_input_is_enough_to_cover_output() {
    let mut tx_builder = setup();
    let tx_out = TxOut::new(50, Script::from_empty());
    tx_builder.add_output(tx_out);

    let tx = tx_builder.build().unwrap();

    assert_eq!(tx.inputs.len(), 1);
    assert_eq!(tx.outputs.len(), 2);
    assert_eq!(tx.outputs[0].value, 50);
}

#[test]
fn test_build_invalid_tx_when_input_is_insufficient_to_cover_output() {
    let mut tx_builder = setup();
    let tx_out = TxOut::new(10000, Script::from_empty());
    tx_builder.add_output(tx_out);

    let tx = tx_builder.build().unwrap();

    assert_eq!(tx.inputs.len(), 5);
    assert_eq!(tx.outputs.len(), 1);
    assert_eq!(tx_builder.input_amount, 500);
    assert_eq!(tx.outputs[0].value, 10000);
}
